### DISCUSSION TITLE
thoughts on `Sendable` and swift 5.10 concurrency warnings?

### DISCUSSION THREAD

**Discussion by jaredh159** [pointfreeco/swift-url-routing#98] [Q&A]
Hi there! Really appreciate the library, and have been using it in production for a couple of years.

I'm working on getting some of my codebases ready for Swift 6 by running 5.10 and turning up as many concurrency warning and strictness settings as I can.

I'm getting a decent number of warnings from this library, stuff like:

```
/my/api/.build/checkouts/swift-url-routing/Sources/URLRouting/Method.swift:27:21: warning: static property 'post' is not concurrency-safe because it is not either conforming to 'Sendable' or isolated to a global actor; this is an error in Swift 6
  public static let post = Self("POST")
```

I would have just opened a PR adding some Sendable conformances, but I'm not really sure how to think about Parser/Printers and Sendability. Are these straightforward to conform to Sendable? or straightforward that they **can not** be conformed to Sendable?  are there plans to do an update addressing concurrency issues? if not, would you guys be willing to give a little guidance on how you'd like to see it done and then accept a PR?

thanks again!

**Comment by mbrandonw** [comment]
Hi @jaredh159, we definitely can (and will) look into updating the library for sendability. For things like the `Method` warning, that is very low hanging fruit and so feel free to PR things like that at anytime.

However, I am curious how exactly you are getting warnings for our library inside your project? When I start a fresh Xcode project with max concurrency warnings, and then `import Parsing`, it builds with no warnings. And I think that's how it's intended to work. Are you doing something special to also get warnings for your dependencies?

**Reply by jaredh159** [reply]
yeah, i'm a bit of an outlier -- i develop almost exclusively in Neovim, and keep a terminal window open with a file watcher running `swift build` or `swift run`. for my server apps i basically never even open xcode, and for my app apps i just use it as a build tool. so, i'm seeing the warnings in the terminal, which maybe surfaces more than the xcode ui does.

the thing about `Method` and sendability is that to make the `static let` not a concurrency warning, i'd have to conform `OneOf` from parser-printers to Sendable, which is what was making me wonder if the conformance was straightforward, since that's a highly generic struct from a different package. or am i missing something dumb?

**Reply by mbrandonw** [reply]
> i develop almost exclusively in Neovim, and keep a terminal window open with a file watcher running `swift build` or `swift run`.

I'm actually still struggling to reproduce this. I created an SPM package, added the library, enabled "StrictConcurrency" in Package.swift, and build from the command line, and there are no warnings. I even added a `var x = 1` to the library to make sure that any warning shows up, and that one does.

Could you create a simple project that shows off this behavior with `swift build` and share it with us? 

> the thing about `Method` and sendability

Oh yes, I didn't look at the code and had thought `Method` was just a simple data type with some statics. I didn't remember it was a parser-printer on its own.

**Reply by jaredh159** [reply]
sure, here's a repro:

https://github.com/jaredh159/url-routing-sendable-repro

in making that, i realized that it's really a transitive dependency through `vapor-routing` (the repro is a nearly empty vapor app). not sure how much that matters, but running `swift build` from the terminal here produces the warning (plus a warning for the `static let router` pattern, which is really the same issue of `OneOf` not being Sendable)

**Reply by mbrandonw** [reply]
Hi @jaredh159, thanks for that. I see now. I think the warning happens if there is a symbol in a dependency with a warning, and then the main library/app target uses that symbol. And I'm guessing doing `@preconcurrency import` doesn't help either does it?

We will look into making this library (and swift-parsing) more sendable friendly soon.
